src/orange/detail/BIHPartitioner.hh

```diff
@@ -0,0 +1,95 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022-2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/detail/BIHPartitioner.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <vector>
+
+#include "orange/BoundingBox.hh"
+#include "orange/OrangeData.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Partition bounding boxes using a surface area heuristic.
+ *
+ * The class take a vector of bounding boxes as an input, and outputs a
+ * Partition object describing the optional partition. To find the optimal
+ * partition, all possible candidate partitions along the x, y, and z axis are
+ * evaluated using a cost function. The cost function is based on a standard
+ * surface area heuristic.
+ */
+class BIHPartitioner
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using VecBBox = std::vector<BoundingBox>;
+    using VecReal3 = std::vector<Real3>;
+    using VecIndices = std::vector<LocalVolumeId>;
+
+    struct Partition
+    {
+        Axis axis = Axis::size_;
+        real_type location = std::numeric_limits<real_type>::infinity();
+
+        VecIndices left_indices;
+        VecIndices right_indices;
+
+        BoundingBox left_bbox;
+        BoundingBox right_bbox;
+
```

Comment:
Maybe these too should use the enum arrays? Then you could index into them using `indices[loc < p.location ? Edge::left : Edge::right]` instead of a bunch of conditional logic.